### PR TITLE
clone script: Allow cloning child jobs as well

### PR DESCRIPTION
**Pull request by ldevulder**
Current version  of `openqa-clone-job` doesn't support multi-machines tests: if you try to clone a HA cluster this will result to only 2 jobs cloned, the supportserver and only a node.

This PR extends `openqa-clone-job` to support multi-machines jobs by adding `--clone-children` option.

Manual tests have been done to ensure that the new version doesn't break anything.

Verification runs:
- jobs cloned with `openqa-clone-job --clone-children --from https://openqa.suse.de 4880442` command: http://1b210.qa.suse.de/tests/7919#dependencies
- jobs cloned with `openqa-clone-job --skip-chained-deps --clone-children --from https://openqa.suse.de 4880442` command: http://1b210.qa.suse.de/tests/7922#dependencies
- jobs cloned with `openqa-clone-job --from https://openqa.suse.de 4880921` command: http://1b210.qa.suse.de/tests/7927#dependencies
- jobs cloned with `openqa-clone-job --skip-chained-deps --from https://openqa.suse.de 4880921`: http://1b210.qa.suse.de/tests/7929

**Note:** there is a CircleCI error on `openqa-clone-job` script that I don't understand... but it seems because `|` character has been replaced by it's hexadecimal code `<7c>`.

### CODE REVIEW

File: lib/OpenQA/Script/CloneJob.pm
Comment (Martchus) [inline lib/OpenQA/Script/CloneJob.pm]: "children" is already plural (of "child"). So please remove the extra "s". And maybe `--clone-children` or `--include-children` would be better.

I'm also wondering why the flag changes this part of the code. Maybe that's something which is useful but alone from the flags name "add-children" is is very unclear and unintuitive why the dependency traversal order would be reversed here.

**Review by Martchus — CHANGES_REQUESTED** [review]
The commit message (title and body) is very misleading. The clone job script already generally supports parallel dependencies and as such MM tests. This PR only covers the case when you not only want to cover the parallel parent but also parallel siblings.

What about

```
clone script: Allow cloning child jobs as well

This is useful to clone full MM clusters.
```

File: script/openqa-clone-job
Comment (Martchus) [inline script/openqa-clone-job:71]:
```suggestion
clone child jobs as well. This is useful if you want to clone a full multi-machine cluster.
```

File: lib/OpenQA/Script/CloneJob.pm
Comment (Martchus) [inline lib/OpenQA/Script/CloneJob.pm:275]:
```suggestion
            # clone child jobs now if needed
```

**Review by Martchus — CHANGES_REQUESTED** [review]
It would also be nice to have some tests for this.

**Comment by Martchus** [comment]
I would also like to note that scheduling jobs with parallel dependencies like this has its caveats: https://progress.opensuse.org/issues/71809

So don't be surprised if not all jobs are assigned completely simultaneously. (There's some code to handle those half-scheduled parallel clusters, though.)

File: lib/OpenQA/Script/CloneJob.pm
Comment (ldevulder) [reply lib/OpenQA/Script/CloneJob.pm]: > I'm also wondering why the flag changes this part of the code. Maybe that's something which is useful but alone from the flags name "add-children" is is very unclear and unintuitive why the dependency traversal order would be reversed here.

What do you mean? The main part of the change is this reverse, as currently if you try to clone a node only the supportserver (the parent job) is cloned, not the other nodes, because these jobs are children of the supportserver. So the idea is to clone the supportserver with this new option to ensure that all dependencies are cloned.
And later in the code we can see another big change: children jobs should be cloned *after* the supportserver, to be sure to have the correct dependencies.

File: script/openqa-clone-job
Comment (ldevulder) [reply script/openqa-clone-job]: Changed, thx!

File: lib/OpenQA/Script/CloneJob.pm
Comment (ldevulder) [reply lib/OpenQA/Script/CloneJob.pm]: Changed.

File: lib/OpenQA/Script/CloneJob.pm
Comment (Martchus) [reply lib/OpenQA/Script/CloneJob.pm]: > What do you mean?

The function you're modifying here is about the asset download of chained dependencies. I'm just not sure how that relates to cloning parallel siblings as well.

> if you try to clone a node only the supportserver (the parent job) is cloned, not the other nodes, because these jobs are children of the supportserver.

I know. The clone script so far only "goes up" the dependency chain. I'm just wondering why this function which is about the asset download would need to be reversed for that.

And in the code down you're also considering children *in addition* to parents. That makes sense because this part of the behavior shouldn't change. However, you're change in this function does *not* consider children in addition so parents are not traversed anymore like before. Maybe that even makes sense as well but then an explaining comment in the code would be helpful.

Comment (ldevulder) [reply lib/OpenQA/Script/CloneJob.pm]: > The function you're modifying here is about the asset download of chained dependencies. I'm just not sure how that relates to cloning parallel siblings as well.

Sorry I thought it was in the other function. So yes you're right I have to double check, maybe this is not needed in fact. I will do tests for this.

Comment (ldevulder) [reply lib/OpenQA/Script/CloneJob.pm]: Yes, it seems that this change is not needed, so I revert it.

**Review by okurz — CHANGES_REQUESTED** [review]
Just a suggestion: I think our tests are pretty well maintained and should help you in development. In general I can recommend a "test first" approach using the already existant automatic tests. We should only merge this PR when we have some test extension to cover the extended/new code but if you struggle to do so yourself we can also provide some help (might take a while though).

The commit message could have a bit more details, e.g. what issue is fixed, why this needs to change, to which versions of which product it applies, link to a bug or a fate entry, the choices you made, etc. Also see https://commit.style/ or http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages. Your PR description already has some details. You should be able to just use that within the git commit message.

File: lib/OpenQA/Script/CloneJob.pm
Comment (okurz) [inline lib/OpenQA/Script/CloneJob.pm:229]: This block becomes a bit too convoluted. I suggest you extract that into a function which you can also get more easily covered in tests. This way you can reduce the indendation level as well.

**Comment by Martchus** [comment]
I'd say adding the support server example to the commit message body should be sufficient.

File: script/openqa-clone-job
Comment (okurz) [inline script/openqa-clone-job:71]: I would appreciate a bit more detail here. Because to my understanding - as you already discussed with mkittler in review comments as well - cloning "full multi-machine clusters" is already available by cloning based on a child. You could explain a bit better that this is related to cloning based on a *parent*, right?

Comment (ldevulder) [reply script/openqa-clone-job:71]: > cloning "full multi-machine clusters" is already available by cloning based on a child

Maybe I misunderstood something or I not explained well: but currently we can't clone a "full multi-machine cluster", only half of it (which is not very useful and don't work ;-))

> You could explain a bit better that this is related to cloning based on a _parent_, right?

Yes the parent ID is passed and cloned, as well as his child. I will add a little more details here.

Comment (Martchus) [reply script/openqa-clone-job:71]: > cloning "full multi-machine clusters" is already available by cloning based on a child

No - that's not true. If one clones a child that covers the parallel parent but not parallel siblings.

It is obvious that with this option one needs to clone the parallel parent instead (otherwise it would be `--clone-siblings`).

File: lib/OpenQA/Script/CloneJob.pm
Comment (ldevulder) [reply lib/OpenQA/Script/CloneJob.pm:229]: Do you want a function for the lines between L223 to L229 or for L222 to L252?

File: script/openqa-clone-job
Comment (ldevulder) [reply script/openqa-clone-job:71]: I added a comment in `usage`.

File: lib/OpenQA/Script/CloneJob.pm
Comment (okurz) [reply lib/OpenQA/Script/CloneJob.pm:229]: I leave it up to you which way you think is more easy to understand and looks prettier :) Obviously, level of indentation is just a superficial symptom of how easy code is to understand or maintain. So functions should have a proper name so that each level of complexity can be read and understood on its own which is easier than trying to understand all levels of complexity/indention here.

File: lib/OpenQA/Script/CloneJob.pm
Comment (alvarocarvajald) [inline lib/OpenQA/Script/CloneJob.pm:238]: Couldn't this be moved outside of the loop? From what I can understand, same results could be reached by doing:
```
push @$child_list, @$parallel if (@$parallel && $job_type eq 'children');
for my $dependencies ($chained, $directly_chained, $parallel) {
    next if (@$dependencies && $dependencies == $parallel && $job_type eq 'children');
    clone_job($_, $options, $clone_map, $depth + 1) for @$dependencies;
}
```
That way we remove a bit of indentation there, and I think the call to `push @$child_list, ...` is clearear.

Or am I missing something?

**Review by alvarocarvajald — APPROVED** [review]
Just a small comment, otherwise LGTM

File: lib/OpenQA/Script/CloneJob.pm
Comment (ldevulder) [reply lib/OpenQA/Script/CloneJob.pm]: Thanks for your help! It works and remove one level of indentation. Added!

File: lib/OpenQA/Script/CloneJob.pm
Comment (ldevulder) [reply lib/OpenQA/Script/CloneJob.pm]: I created a function `get_deps` to clear the code a little bit.

File: script/openqa-clone-job
Comment (Martchus) [inline script/openqa-clone-job:71]:
```suggestion
clone child jobs as well. This is useful if you want to clone a full multi-machine cluster.
```

**Review by Martchus — COMMENTED** [review]
Looks good, I only found a typo. Not sure how much effort unit tests would be at this point but of course having them would be a plus.

File: script/openqa-clone-job
Comment (ldevulder) [reply script/openqa-clone-job]: Fixed, thx.

File: t/35-script_clone_job.t
Comment (Martchus) [inline t/35-script_clone_job.t:161]: Please remove these leftovers.

File: t/35-script_clone_job.t
Comment (Martchus) [inline t/35-script_clone_job.t:183]: I don't think the tidy check is happy with these. It would also make sense to write these hashes in a more compact way.

File: t/35-script_clone_job.t
Comment (Martchus) [inline t/35-script_clone_job.t:62]:
```suggestion
```

File: t/35-script_clone_job.t
Comment (ldevulder) [reply t/35-script_clone_job.t:161]: Yes of course, it's just a first push to be able to run the `.t` test, as I can't run it locally.

File: lib/OpenQA/Script/CloneJob.pm
Comment (Martchus) [inline lib/OpenQA/Script/CloneJob.pm:235]:
```suggestion
    my @job_types = ('parents');
```

File: t/35-script_clone_job.t
Comment (Martchus) [inline t/35-script_clone_job.t:160]:
```suggestion
        settings => \%node1_settings,
```
or, if you really need a copy (likely not):
```suggestion
        settings => [%node1_settings],
```
(The same counts for the other occurrences.)

File: lib/OpenQA/Script/CloneJob.pm
Comment (Martchus) [inline lib/OpenQA/Script/CloneJob.pm:223]:
```suggestion
    return ($chained // [], $directly_chained // [], $parallel // []);
```

**Review by Martchus — CHANGES_REQUESTED** [review]

File: t/35-script_clone_job.t
Comment (Martchus) [reply t/35-script_clone_job.t]: To run the test locally, use `t/test_postgresql /dev/shm/tpg` and then `TEST_PG='DBI:Pg:dbname=openqa_test;host=/dev/shm/tpg' OPENQA_BASEDIR= prove -v t/35-script_clone_job.t` (from https://open.qa/docs/#testing). See https://github.com/Martchus/openQA-helper for an example of a good development environment.

Comment (okurz) [reply t/35-script_clone_job.t]: the documentation already shows a simpler example:

```
make test KEEP_DB=1 TESTS=t/35-script_clone_job.t
```

but your example also works showing the low level steps in detail.

File: lib/OpenQA/Script/CloneJob.pm
Comment (ldevulder) [reply lib/OpenQA/Script/CloneJob.pm:223]: I simply copy/paste without trying to improve, thx for getting this!

File: lib/OpenQA/Script/CloneJob.pm
Comment (ldevulder) [reply lib/OpenQA/Script/CloneJob.pm]: Fixed!

File: t/35-script_clone_job.t
Comment (ldevulder) [reply t/35-script_clone_job.t]: Fixed!

File: t/35-script_clone_job.t
Comment (ldevulder) [reply t/35-script_clone_job.t]: Ok, I will try to compact it as soon as I have a test that works.

**Comment by ldevulder** [comment]
@Martchus I have this error in `c/circle: t` and I don't understand it:
```
The test suite died before a plan was produced. You need to have a plan.
No plan
```

Any hints?

**Comment by Martchus** [comment]
If you switch to the "Steps" tab you'll see more helpful output:

```
[09:16:23] t/35-script_clone_job.t ................................... Bareword found where operator expected at t/35-script_clone_job.t line 160, near "%%node1_settings"
	(Missing operator before node1_settings?)
Bareword found where operator expected at t/35-script_clone_job.t line 167, near "%%node2_settings"
	(Missing operator before node2_settings?)
syntax error at t/35-script_clone_job.t line 160, near "%%node1_settings"
syntax error at t/35-script_clone_job.t line 167, near "%%node2_settings"
Execution of t/35-script_clone_job.t aborted due to compilation errors.
```

Otherwise, the logs within the "Artifacts" tab might help as well. However, this problem should actually vanish since you've implemented my review comments.

Note that I had to go back to the previous test run because the CI has been retriggered. Luckily CircleCI preserves the history of CI pipelines for a PR: https://app.circleci.com/pipelines/github/os-autoinst/openQA?branch=pull%2F3487

I hope you've seen my comment about running tests locally: https://github.com/os-autoinst/openQA/pull/3487#discussion_r514973636  
For faster development it really makes more sense to run tests locally.

File: t/35-script_clone_job.t
Comment (Martchus) [inline t/35-script_clone_job.t:164]:
```suggestion
        parents => {
            Parallel => [7934],
        },
```

The same counts for the other occurrences. Note that `()` is used in Perl to create a "list" (not sure about the terminology). From a "list" other stuff like hashes and arrays are initialized. However, if you create nested structures, "lists" are just "unpacked"/"expanded" (still not sure about the terminology). So `(…, parents => (Parallel => (7834)))` is actually equivalent to just `(…, parents => Parallel => 7834)` which is not what you want. So you need to use `{}` and `[]` to create a hash/array-ref explicitly.

**Review by Martchus — CHANGES_REQUESTED** [review]

File: t/35-script_clone_job.t
Comment (ldevulder) [reply t/35-script_clone_job.t]: Changed, thx!

File: t/35-script_clone_job.t
Comment (ldevulder) [reply t/35-script_clone_job.t]: With the latest change (thanks again!) tidy looks happier.

**Comment by ldevulder** [comment]
@Martchus the latest version looks good, the test code for `get_deps` function seems to work well.
I first though to create a test for `clone_job` function, but this one seems more complicated to test (maybe that's why it is not done yet ;-)).

Let me know if it's OK for you.

**Comment by okurz** [comment]
@ldevulder

* regarding "tidy" make sure you are running perl-Perl-Tidy-20201001 as specified in https://github.com/os-autoinst/openQA/blob/master/dependencies.yaml#L74 and then call `tools/tidy`
* OBS checks are currently running and waiting for package dependencies to be built so nothing to be done on your side except for waiting. Be aware that there can be false reports from "failed package builds" which can be "unresolvable" as long as the main package is actually not yet finished building. So we only need to look into failures in case when all packages have actually finished building.

**Review by okurz — APPROVED** [review]
But overall this looks very nice. Thanks so much for your contribution so far! Feel free to ask us to take over for fixing the tidy issues or something to get this merged.

**Review by Martchus — COMMENTED** [review]
Does this work for (directly) chained dependencies as well? Judging by the code I'm not sure.

**Comment by ldevulder** [comment]
> Does this work for (directly) chained dependencies as well? Judging by the code I'm not sure.

What do you mean, do you have an example to check? If it was working previously it still should be (at least my regression tests look good).

**Comment by ldevulder** [comment]
> * regarding "tidy" make sure you are running perl-Perl-Tidy-20201001 as specified in https://github.com/os-autoinst/openQA/blob/master/dependencies.yaml#L74 and then call `tools/tidy`

Thx, `checkstyle` is OK now :)

**Review by Martchus — APPROVED** [review]
The new parameter is called `--clone-children` which implies that it affects all dependency types and that might actually be useful. However, the code looks like that it is only implementing the new behavior for parallel dependencies. At least the lines `$job->{settings}->{_PARALLEL_JOBS}    = join(',', @new_parallel) if @new_parallel && !defined $child_list;` and `push @$child_list, @$parallel if (@$parallel && $job_type eq 'children');` seem specific to parallel dependencies. But I've just tested it myself. I've cloned https://openqa.opensuse.org/tests/1458924#dependencies with your version and all chained dependencies are correctly recursively cloned (and without the flag the behavior is unchanged). So everything is fine.

**Comment by ldevulder** [comment]
There is now issues with Codecov, but I don't know what to do here...

**Comment by Martchus** [comment]
The coverage changes seem unrelated. However, it seems plausible that not all of the code within `sub clone_job` is covered because your test only calls `get_deps`. I guess it is ok, though. We've tested it manually and I wouldn't require you to add tests for every part where we're currently lacking test coverage.

**Comment by okurz** [comment]
Yeah, the check "codecov/patch" means that the code that you changed is not fully covered by tests. This of course could have been prevented by having full code statement coverage in before so that your PR is only changing covered code. The alternative is that of course you add the corresponding test code within your PR which would be great but also not mandatory as Martchus has written

**Comment by okurz** [comment]
The failing OBS checks are related to https://progress.opensuse.org/issues/71857 which is already "Urgent", so I will force-merge here.

**Comment by ldevulder** [comment]
> The alternative is that of course you add the corresponding test code within your PR which would be great but also not mandatory as Martchus has written

Which part are missing? I could try to add some tests in my "spare" time (so between two things at work ;-)).

**Comment by okurz** [comment]
The codecov report provide a good start if you prefer something visual. E.g. https://codecov.io/gh/os-autoinst/openQA/pull/3487/diff?src=pr&el=tree#diff-bGliL09wZW5RQS9TY3JpcHQvQ2xvbmVKb2IucG0= shows in red all code blocks that are not yet executed in tests.

To gather code coverage locally http://open.qa/docs/#testing also has some hints, e.g. try

```
make coverage TESTS=t/35-script_clone_job.t
```

and you should see a summary of coverage and a link to a locally generated HTML report with the coverage. Then try to add test code according to ensure the uncovered code blocks are called and crosscheck by calling the above command again to see if the coverage increases.

**Pull request merged**
